### ISSUE TITLE
routes not being set correctly when switching to cross-subnet ip encapsulation

### ISSUE THREAD

**Issue by bradbehle** [projectcalico/calico#1325]
We have a cluster we have upgraded to calico 2.6.1, and at that time also enabled cross-subnet IP in IP encapsulation (so pod to pod traffic b/t workers in the same subnet aren't encapsulated).  All the workers are in the same subnet, but it doesn't appear calico has updated all the routes, so some of them still use the tunl0 interface.  And worse, in some cases node A uses it to send to node B, but node B doesn't use it to send back to node A.  So pod to pod traffic between some nodes is blocked.

## Expected Behavior
I would expect, and have seen in other clusters, that when cross-subnet IPinIP is turned on, the routes are immediately updated on all nodes to correctly route traffic.

## Current Behavior
Here is some output that shows our cluster calico configuration, and the routes that don't seem to be getting set correctly:

```
calicoctl get nodes -o wide --config calicoctl.cfg 
NAME                                                          ASN       IPV4               IPV6   
kube-dal12-cr773988613fbe4fa7bb75ee537d9ba5a7-w10.cloud.ibm   (64512)   10.184.105.18/26          
kube-dal12-cr773988613fbe4fa7bb75ee537d9ba5a7-w11.cloud.ibm   (64512)   10.184.105.35/26          
kube-dal12-cr773988613fbe4fa7bb75ee537d9ba5a7-w12.cloud.ibm   (64512)   10.184.105.17/26          
kube-dal12-cr773988613fbe4fa7bb75ee537d9ba5a7-w13.cloud.ibm   (64512)   10.184.105.10/26          
kube-dal12-cr773988613fbe4fa7bb75ee537d9ba5a7-w14.cloud.ibm   (64512)   10.184.105.24/26          
kube-dal12-cr773988613fbe4fa7bb75ee537d9ba5a7-w15.cloud.ibm   (64512)   10.184.105.53/26          
kube-dal12-cr773988613fbe4fa7bb75ee537d9ba5a7-w16.cloud.ibm   (64512)   10.184.105.23/26          
kube-dal12-cr773988613fbe4fa7bb75ee537d9ba5a7-w17.cloud.ibm   (64512)   10.184.105.27/26          
kube-dal12-cr773988613fbe4fa7bb75ee537d9ba5a7-w18.cloud.ibm   (64512)   10.184.105.47/26          

calicoctl get ippool -o yaml --config calicoctl.cfg 
- apiVersion: v1
  kind: ipPool
  metadata:
    cidr: 172.30.0.0/16
  spec:
    ipip:
      enabled: true
      mode: cross-subnet
    nat-outgoing: true


10.184.105.27 route -n:
172.30.0.0      10.184.105.17   255.255.255.192 UG    0      0        0 tunl0
172.30.12.64    10.184.105.35   255.255.255.192 UG    0      0        0 tunl0
172.30.59.128   10.184.105.23   255.255.255.192 UG    0      0        0 eth0
172.30.99.128   10.184.105.24   255.255.255.192 UG    0      0        0 tunl0
172.30.155.192  10.184.105.18   255.255.255.192 UG    0      0        0 eth0
172.30.221.128  10.184.105.53   255.255.255.192 UG    0      0        0 tunl0
172.30.224.128  10.184.105.10   255.255.255.192 UG    0      0        0 tunl0

10.184.105.18 route -n:

172.30.0.0      10.184.105.17   255.255.255.192 UG    0      0        0 tunl0
172.30.12.64    10.184.105.35   255.255.255.192 UG    0      0        0 tunl0
172.30.59.128   10.184.105.23   255.255.255.192 UG    0      0        0 tunl0
172.30.99.128   10.184.105.24   255.255.255.192 UG    0      0        0 tunl0
172.30.221.128  10.184.105.53   255.255.255.192 UG    0      0        0 tunl0
172.30.224.128  10.184.105.10   255.255.255.192 UG    0      0        0 tunl0
172.30.228.192  10.184.105.27   255.255.255.192 UG    0      0        0 tunl0
172.30.232.128  10.184.105.47   255.255.255.192 UG    0      0        0 tunl0

10.184.105.35 route -n:

172.30.0.0      10.184.105.17   255.255.255.192 UG    0      0        0 eth0
172.30.59.128   10.184.105.23   255.255.255.192 UG    0      0        0 eth0
172.30.99.128   10.184.105.24   255.255.255.192 UG    0      0        0 eth0
172.30.155.192  10.184.105.18   255.255.255.192 UG    0      0        0 eth0
172.30.221.128  10.184.105.53   255.255.255.192 UG    0      0        0 eth0
172.30.224.128  10.184.105.10   255.255.255.192 UG    0      0        0 eth0
172.30.228.192  10.184.105.27   255.255.255.192 UG    0      0        0 eth0
172.30.232.128  10.184.105.47   255.255.255.192 UG    0      0        0 eth0
```

## Your Environment

* Calico version: 2.6.1
* Orchestrator version: kubernetes 1.7.4
* Operating System and version: Ubuntu 16.04

**Comment by robbrockbank** [comment]
Hi @bradbehle 

Without looking at the logs, I'm wondering if confd has skipped an update on the BGP configuration for some reason.  I have worked in this area recently and shored up error handling in the master confd that could potentially resolve this.

As a first step, could you reapply the IPPool configuration and see if that triggers everything into life?

**Comment by bradbehle** [comment]
FYI, restarting calico-node did not help.  But before I saw your update above I went to each node and manually deleted (using `sudo route -del ...`) the routes that were mistakenly using tunl0, and those routes were then immediately replaced with the correct routes (routing directly over eth0).  Does that make it more or less likely that it is a confd issue?

I will try to recreate this problem, since it could cause serious issues for us as we move to using cross-subnet IP in IP.  If I can recreate I'll try to reapply the ippool to see if that helps, and will post more info here.

**Comment by robbrockbank** [comment]
Ah, I would have expected a restart of calico/node to re-render the templates too.

So, perhaps there is some configuration issue, or other error that is causing this.  If you see this again, perhaps you could grab the logs from the affected node(s).  Feel free to DM me on slack if you'd prefer to send your diags directly to me.

**Comment by bradbehle** [comment]
I was able to find this same situation (tunl0 routes that aren't getting replaced on somenodes when cross-subnet is enabled) on a different cluster.  I have sent the logs and cluster config to @robbrockbank

**Comment by robbrockbank** [comment]
I'm seeing the following logs from bird:

`bird: filters, line 18: ~ applied on unknown type pair`

Which was an issue addressed in the PR #1218 - back porting to v2.6 in #1354

**Comment by robbrockbank** [comment]
didn't mean to close - apologies.

**Comment by fasaxc** [comment]
Fix will be released in Calico v2.6.3, to be released shortly.
